edu-services/cm-service/cm-impl/hibernate-impl/impl/src/java/org/sakaiproject/coursemanagement/impl/CourseSitePublishServiceImpl.java

```diff
@@ -0,0 +1,244 @@
+    *             if this method is invoked at any time during this
+    *             time period, unpublished course sites associated
+    *             with this term will be published.
+    *
+    * </p>
+    * <p>
+    * There is a special use case to consider.
+    * Say that this method is invoked, and all the course sites for the current term are published.
+    * Now, suppose one particular instructor forgot to add some content, or made some mistakes setting up his course site, and he wants to unpublish his couse site while
+    * he makes some changes to it.  He goes into sakai, and uses the site setup tool to mark the course site as unpublished.  But, suppose the next day this service is
+    * invoked again (via a quartz job).  His course would get published again, even though he doesn't want it to be, and has manually gone into sakai and set it to
+    * unpublished.  In order to prevent this case from happening, whenever this service marks a course site as published, it also sets a site property called <i>course_site_publish_service.publish.set</i>.
+    * Then, if an instructor manually resets his site to be unpublished, this service will not automatically publish it.  Rather, the instructor who manually unpublished the
+    * course site will have to go into sakai, use the site setup tool, and manually publish it.
+    * </p>
+    * </br></br>
+    * @param numDaysBeforeTermStarts   number of days before a term starts that course sites should be published.
+    * </br></br>
+    * @return the number of course sites that were published.
+    */
+   public int publishCourseSites(int numDaysBeforeTermStarts) {
+
+      logger.info("publishCourseSites(" + numDaysBeforeTermStarts + " days before the term starts)");
+
+      Date today             = new Date();
+      Date publishDate       = null;
+      int  numSitesPublished = 0;
+
+      try {
+         // get the list of the academic terms
+         List<AcademicSession> academicSessions = courseManagementService.getAcademicSessions();
+
+         for(AcademicSession academicSession : academicSessions) {
+            publishDate = new Date(academicSession.getStartDate().getTime() - numDaysBeforeTermStarts * ONE_DAY_IN_MS);
+            // see if the academic is scheduled to start within the specified number of days
+            if (publishDate.getTime() < today.getTime() && today.getTime() < academicSession.getEndDate().getTime()) {
+
+               // get a list of all published and unpublished course sites in ascending creation date order which are associated with the specified academic session
+               Hashtable<String, String> propertyCriteria = new Hashtable<String, String>();
+               propertyCriteria.put("term_eid", academicSession.getEid());
+               List<Site> sites = (List<Site>)siteService.getSites(SelectionType.NON_USER, "course", null, propertyCriteria, SortType.CREATED_ON_ASC, null);
```

Comment:
Looks like we're iterating over every site in the current academic session(s).
Certainly not the only cron job that does this, so it's probably okay, but I'd be in favor of cutting this down to return only unpublished sites.
Doesn't look like this can currently be done in the SiteService API though.
Perhaps we could open a jira ticket for this functionality?